src/orion/core/worker/trial.py

```diff
@@ -445,3 +445,36 @@ def compute_trial_hash(
         return hashlib.md5(
             (params + experiment_repr + lie_repr).encode("utf-8")
         ).hexdigest()
+
+
+class TrialCM:
+    __slots__ = ("_cm_experiment", "_cm_trial")
+
+    def __init__(self, experiment, trial):
+        self._cm_experiment = experiment
+        self._cm_trial = trial
+
+    def __getattribute__(self, name):
+        if name in {"_cm_experiment", "_cm_trial", "__enter__", "__exit__"}:
+            return object.__getattribute__(self, name)
+        return getattr(self._cm_trial, name)
+
+    def __setattr__(self, name, value):
+        if name not in {"_cm_experiment", "_cm_trial"}:
+            setattr(self._cm_trial, name, value)
+        else:
+            object.__setattr__(self, name, value)
+
+    def __enter__(self):
+        return self._cm_trial
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        try:
```

Comment:
If the trial status is not `reserved` locally, the release may _corrupt_ the database. For instance if `experiment.observe(trial)` is called before the release, if will change status from 'completed' to 'broken'. To be clear, this is not due to this new implementation but rather was an existing bug that I just found out.

I would suggest adding `was` argument to `set_trial_status` in `storage` ([here for example](https://github.com/Epistimio/orion/blob/develop/src/orion/storage/base.py#L330) to specify what is supposed to be the current status. By default it would be set to None and would use trial.status in the `where` query as is currently the case.

In `ExperimentClient.release`, `set_trial_status(trial, status)` would be changed to 
`set_trial_status(trial, status, was='reserved')`

This would require no changes from this PR implementation of `__exit__`